crate/server/src/core/operations/validate.rs

```diff
@@ -0,0 +1,586 @@
+        certificate_by_uid(unique_identifier, kms, user, params).await
+    }))
+    .await;
+
+    // filtering errors
+    let is_err = res.iter().all(|x| x.is_err());
+
+    // checking if there are any errors
+    if !is_err {
+        KResult::Ok(
+            res.iter()
+                .map(|x| x.clone().expect("safe unwrap"))
+                .collect(),
+        )
+    } else {
+        KResult::Err(KmsError::from(KmipError::ObjectNotFound(
+            "There is a Kms Certificate UID that cannot be retrieved".to_string(),
+        )))
+    }
+}
+
+// Fetches a certificate. If it fails, returns the according error
+async fn certificate_by_uid(
+    unique_identifier: &str,
+    kms: &KMS,
+    user: &str,
+    params: Option<&ExtraDatabaseParams>,
+) -> KResult<Vec<u8>> {
+    let uid_own = retrieve_object_for_operation(
+        unique_identifier,
+        ObjectOperationType::Validate,
+        kms,
+        user,
+        params,
+    )
+    .await;
+    match uid_own {
+        Err(e) => KResult::Err(e),
+        Ok(kms_object) => {
+            if let Object::Certificate {
+                certificate_type: _,
+                certificate_value,
+            } = kms_object.object
+            {
+                KResult::Ok(certificate_value)
+            } else {
+                KResult::Err(KmsError::from(KmipError::InvalidKmipObject(
+                    cosmian_kmip::kmip::kmip_operations::ErrorReason::Invalid_Object_Type,
+                    String::from("Requested a Certificate Object, got a ")
+                        + &kms_object.object.object_type().to_string(),
+                )))
+            }
+        }
+    }
+}
+
+fn validate_chain_date(
+    certificates: &mut [Vec<u8>],
+    date: Option<String>,
+) -> KResult<ValidityIndicator> {
+    println!("Looking for a date to compare with");
+    let current_date = {
+        if let Some(date) = date {
+            println!("Converting date from input");
+            Asn1Time::from_str(date.as_str())
+        } else {
+            println!("Getting current date");
+            Asn1Time::days_from_now(0)
+        }
+    }?;
+    println!("Got a date!");
+    certificates
+        .iter()
+        .try_fold(ValidityIndicator::Valid, |acc, certificate| {
+            let validation = validate_date(certificate, &current_date)?;
+            KResult::Ok(acc.and(validation))
+        })
+}
+
+fn validate_date(certificate: &[u8], date: &Asn1Time) -> KResult<ValidityIndicator> {
+    let certificate = X509::from_der(certificate)?;
+    let (certificate_validity_start, certificate_validity_end) =
+        (certificate.not_before(), certificate.not_after());
+    if date.as_ref() <= certificate_validity_end && certificate_validity_start <= date.as_ref() {
+        println!("DATE IS VALID");
+        KResult::Ok(ValidityIndicator::Valid)
+    } else {
+        println!("DATE IS INVALID");
+        KResult::Ok(ValidityIndicator::Invalid)
+    }
+}
```

Comment:
I would argue this is a simplified version:
```suggestion
fn validate_date(certificate: &X509, date: &Asn1Time) -> ValidityIndicator {
    let start = certificate.not_before();
    let now = date.as_ref();
    let stop = certificate.not_after();
    if start <= now && now <= stop {
        ValidityIndicator::Valid
    } else {
        ValidityIndicator::Invalid
    }
}
```